### PR TITLE
ci(rust-ci): publish docs to GitHub Pages

### PR DESCRIPTION
**Pull request by shenanigansd**
Part of the efforts to get documentation going. It's just generated docs of the internal API, but it gets us an already automated jumping off point.

### CODE REVIEW

File: .github/workflows/rust-ci.yaml [hunk 1]
```diff
@@ -128,3 +128,55 @@ jobs:
           YARA_INCLUDE_DIR: ${{ github.workspace }}/.yara/${{ needs.yara-test.outputs.include-dir }}
           YARA_LIBRARY_PATH: ${{ github.workspace }}/.yara/${{ needs.yara-test.outputs.library-path }}
         run: cargo test --no-fail-fast
+
+  docs-build:
+    needs: [yara]
+    runs-on: ubuntu-latest
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:134]: As with the other jobs,
```suggestion
    runs-on: ubuntu-22.04
```

File: .github/workflows/rust-ci.yaml [hunk 2]
```diff
@@ -128,3 +128,55 @@ jobs:
...
+      - name: Download YARA
+        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        with:
+          name: ${{ needs.yara.outputs.artifacts }}
+          path: .yara
+
+      - name: "Build docs with cargo"
+        run: cargo doc --no-deps --document-private-items
+        env:
+          RUST_BACKTRACE: full
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:154]: Do we need this now that we've troubleshot and fixed the job failures?

File: .github/workflows/rust-ci.yaml [hunk 3]
```diff
@@ -128,3 +128,55 @@ jobs:
...
+          YARA_INCLUDE_DIR: ${{ github.workspace }}/.yara/${{ needs.yara.outputs.include-dir }}
+          YARA_LIBRARY_PATH: ${{ github.workspace }}/.yara/${{ needs.yara.outputs.library-path }}
+
+      - name: "Upload artifact"
+        uses: actions/upload-pages-artifact@a753861a5debcf57bf8b404356158c8e1e33150c # v2.0.0
+        with:
+          path: ./target/doc
+
+  docs-deploy:
+    if: ${{ github.ref == 'refs/heads/main' }}
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:164]:
```suggestion
    if: github.ref == 'refs/heads/main'
```

File: .github/workflows/rust-ci.yaml [hunk 4]
```diff
@@ -128,3 +128,55 @@ jobs:
...
+
+    # Sets permissions of the GITHUB_TOKEN to allow deployment to GitHub Pages
+    permissions:
+      contents: read
+      pages: write
+      id-token: write
+
+    environment:
+      name: github-pages
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:173]: To be clear about what's deployed on the environment.
```suggestion
      name: docs
```

File: .github/workflows/rust-ci.yaml [hunk 5]
```diff
@@ -128,3 +128,55 @@ jobs:
...
+      url: ${{ steps.deployment.outputs.page_url }}
+
+    runs-on: ubuntu-latest
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:176]:
```suggestion
    runs-on: ubuntu-22.04
```

File: .github/workflows/rust-ci.yaml [hunk 6]
```diff
@@ -128,3 +128,55 @@ jobs:
...
+
+    needs: docs-build
```
Comment (sid-maddy) [inline .github/workflows/rust-ci.yaml:178]: nit: Move `needs` near the top of the job definition, just below the `if`.

**Review by sid-maddy — CHANGES_REQUESTED** [review]

File: .github/workflows/rust-ci.yaml (hunk 6 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml]: Agreed
74adcab44f341f450101dba01883cb2c2bb35fa5

File: .github/workflows/rust-ci.yaml (hunk 5 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml]: 99e2331ab6dce0b6484eaaba36508d91aa9743b0

File: .github/workflows/rust-ci.yaml (hunk 1 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml]: 99e2331ab6dce0b6484eaaba36508d91aa9743b0

File: .github/workflows/rust-ci.yaml (hunk 4 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml:175]: Will this work? I thought that specific name was required for this to work.
We can give it a try.

File: .github/workflows/rust-ci.yaml (hunk 3 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml:166]: What's the difference between these two?

File: .github/workflows/rust-ci.yaml (hunk 2 above)
Comment (shenanigansd) [reply .github/workflows/rust-ci.yaml]: Doesn't really matter, doesn't hurt anything, but it doesn't effect anything either, so we can remove it.
25dbdb3fa43cbfce39c8e4e12258503129710556

**Review by sid-maddy — APPROVED** [review]
:shipit:

**Pull request merged**